### ISSUE TITLE
Segmentation fault when starting neo-cli on Linux

### ISSUE THREAD

**Issue by Crypto2** [neo-project/neo-cli#295]
Using neo-cli with application logging enabled (dotnet neo-cli.dll --rpc --log) I keep getting a Segmentation fault during startup. I then have to delete my Chain/Index/ApplicationLog folders and resync with chain.acc.zip in order to start NEO up again.

**Comment by Crypto2** [comment]
Any update on this. Even when it is running it's running really poorly compared to older versions, being slow and RPC requests timing out with no reply. Not sure if this is just because of --log which we just had to add recently or what???

**Comment by vncoelho** [comment]
Hi, @Crypto2.

Trying to solve this issue:

Apparently, it something related to the plugins configurations.
The error that I receive is:

```cs
[ERROR][02/09/2019 22:28:14][Thread 0003][akka://NeoSystem/user/$a] Object reference not set to an instance of an object.
Cause: System.NullReferenceException: Object reference not set to an instance of an object.
   at Neo.Plugins.SimplePolicyPlugin.get_MaxTxPerBlock()
   at Neo.Ledger.MemoryPool.LoadMaxTxLimitsFromPolicyPlugins()
   at Neo.Ledger.MemoryPool.UpdatePoolForBlockPersisted(Block block, Snapshot snapshot)
   at Neo.Ledger.Blockchain.OnPersistCompleted(Block block)
   at Neo.Ledger.Blockchain.Persist(Block block)
   at Neo.Ledger.Blockchain.OnImport(IEnumerable`1 blocks)
   at Akka.Actor.UntypedActor.Receive(Object message)
   at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
   at Akka.Actor.ActorCell.ReceiveMessage(Object message)
   at Akka.Actor.ActorCell.Invoke(Envelope envelope)
[ERROR][02/09/2019 22:28:14][Thread 0004][akka://NeoSystem/user/$a] Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Persistence.LevelDB.LevelDBStore)
Cause: [akka://NeoSystem/user/$a#184477984]: Akka.Actor.PostRestartException: Exception post restart (System.NullReferenceException) ---> System.TypeLoadException: Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Persistence.LevelDB.LevelDBStore) ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Neo.Ledger.Blockchain..ctor(NeoSystem system, Store store)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at Akka.Actor.Props.ActivatorProducer.Produce()
   at Akka.Actor.Props.NewActor()
   --- End of inner exception stack trace ---
   at Akka.Actor.Props.NewActor()
   at Akka.Actor.ActorCell.CreateNewActorInstance()
   at Akka.Actor.ActorCell.<>c__DisplayClass109_0.<NewActor>b__0()
   at Akka.Actor.ActorCell.UseThreadContext(Action action)
   at Akka.Actor.ActorCell.NewActor()
   at Akka.Actor.ActorCell.FinishRecreate(Exception cause, ActorBase failedActor)
   --- End of inner exception stack trace ---
```

**Comment by jsolman** [comment]
That could happen if you are using master of Neo GitHub and trying to use the 2.9.4 released plugin of SimplePlugin. It needs this Commit that hasn’t been included in a release yet to avoid the error above:
https://github.com/neo-project/neo-plugins/commit/a7fd6590cea4530d2e2082aec0a1a6ea5bdd6451

**Comment by vncoelho** [comment]
Perhaps was it, Jeff, I tested with last masters as well, but sometimes we get confused with all different versions.
I will double check.

**Comment by vncoelho** [comment]
![image](https://user-images.githubusercontent.com/17395762/52527548-6cefd000-2cb1-11e9-9f16-c4efd01c4b05.png)

Checked again. Not sure yet, even with last commits this errors persisted.
Let leave this thread open until further investigation.

**Comment by vncoelho** [comment]
I just had another problem, after closing the neo-cli and reopening it gave these errors. Thus, it might be also related to something when restarting a node.

**Comment by vncoelho** [comment]
@jsolman, the strange think is that sometimes it syncs without problems and sometimes it crashes...

**Comment by Crypto2** [comment]
I was using the 2.9.4 tagged release and 2.9.4 plugins. It didn't print out any error messages just Segmentation fault. I've gone back to 2.9.2 for now since it runs better. 2.9.4 even when running was so laggy/slow to be unusable :(

**Comment by jsolman** [comment]
@Crypto2 were you using the setting in 2.9.4 to limit your connection count?

**Comment by jsolman** [comment]
I have a PR out for a more graceful shutdown but I need to do a little more testing on it and try one more thing before it will be merged.

**Comment by Crypto2** [comment]
@jsolman Nope, the only non-default item in my config is the RPC bind address and only command line parms are --rpc and --log

**Comment by Crypto2** [comment]
Is this and the slow running fixed in 2.10.0?

**Comment by jsolman** [comment]
Try it and let me know. I don’t usually use the application log plugin. I am not getting the segmentation fault or performance issue.

**Comment by Crypto2** [comment]
You do know the application log plugin is essentially required to support token deposits (at least as far as I can tell, docs aren't exactly great here.) You'd think it would be well tested by the core devs and major bugs like this fixed ASAP. Unfortunately I have to try 2.10.0 now since 2.9.2 won't sync past 3473781 now. :(

**Comment by jsolman** [comment]
I cannot speak for all the other core devs. Personally, I have tracked NEP5 balances using my own code for the 9 months rather than the application log. I also wrote an efficient way to track the NEP5 balances as a Neo plugin that works with 2.10.0.

https://github.com/neo-project/neo-plugins/tree/master/RpcNep5Tracker

**Comment by jsolman** [comment]
> You'd think it would be well tested by the core devs and major bugs like this fixed ASAP.

The plugin should be working. Please try it with the 2.10.0 release. If you have a problem report details of how to reproduce it and I will look into it.

**Comment by jsolman** [comment]
Were the problems you were having related to syncing using the ImportBlocks plugin? I found an issue with the way the plugins load where it could start importing blocks potentially before the the plugin loaded; a fix for that will be in 2.10.1.

**Comment by jsolman** [comment]
I just took a look at the plugin source again, and I see something I can do to improve its performance. Also, I see an issue with it if `neo-cli` is killed at an inopportune time since it isn’t doing its writes with batching. I’ll make a PR to improve the plugin.

**Comment by Crypto2** [comment]
We did sync using ImportBlocks but not sure if it caused the problem or not, just once the program was closed we'd get a segfault on the next start unless we deleted the Chain_* folder and resynced. Since we have so many addresses we end up walking the blockchain for deposit processing (which is how a lot of our coins end up working so don't mind just need working wallet that doesn't need to resync every time you restart it.)

We do always close the wallet by typing 'exit', never kill or kill -9 or anything like that.

**Comment by jsolman** [comment]
The wallet does not stop properly when typing exit when using the `ImportBlocks` plugin. I am planning to fix this.

**Comment by jsolman** [comment]
I opened an issue to track the exit issue while syncing blocks using the `ImportBlocks` plugin: https://github.com/neo-project/neo-plugins/issues/59

**Comment by jsolman** [comment]
I opened an issue to track application log plugin improvement also: https://github.com/neo-project/neo-plugins/issues/60

**Comment by jsolman** [comment]
> just once the program was closed we'd get a segfault on the next start unless we deleted the Chain_* folder

What version of leveldb were you using for my reference? Also,  what OS were you using?

I will make changes to the application log plugin today which should hopefully resolve your problems.

**Comment by Crypto2** [comment]
Using Ubuntu 16.04 LTS (since it says newer versions aren't supported) using the stock leveldb 1.18-5

**Comment by jsolman** [comment]
I've started a PR for batching writes in the Application Log plugin: neo-project/neo-plugins/pull/62

**Comment by jsolman** [comment]
I fixed the ImportBlocks plugin. I have a PR out here: neo-project/neo-plugins/pull/63

**Comment by jsolman** [comment]
With these 2 PRs, this issue should be resolved.

**Comment by vncoelho** [comment]
Are you getting this error @Crypto2?

```cs
Unhandled Exception: System.ArgumentException: Unable to resolve the target Provider
Parameter name: self
   at Akka.Actor.Futures.Ask[T](ICanTell self, Func`2 messageFactory, Nullable`1 timeout, CancellationToken cancellationToken)
   at Neo.Plugins.ImportBlocks.OnImport()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
/opt/start_node.sh: line 14:    38 Aborted                 (core dumped) dotnet neo-cli.dll --rpc --log
```

**Comment by vncoelho** [comment]
I will soon test with your new PR, @jsolman.

**Comment by superboyiii** [comment]
@vncoelho Crypto2's issue is like this, not that one. Yours seems like the param [self] in settings.cs  hasn't been set into ImportBlocks.cs.
![image](https://user-images.githubusercontent.com/38493437/55308065-12424b80-548c-11e9-969a-243007876166.png)

**Comment by shoxive** [comment]
> ```cs
> previous
> ```
Hi,I met this error after finish downloading the offline package. Any suggest of this problem?

**Comment by shoxive** [comment]
@vncoelho 
Hi,I met this error after finish downloading the Full offline package. Any suggest of this problem(It looks like ImportBlocks error)?

**Comment by vncoelho** [comment]
Hi @shoxive, thanks for replying it.

Could you please print the message here and, if possible, try with 2.10.2? :)
Thanks,

Since it is not a critical issue we are not really focusing on it. But let's try to cover all these exceptions.

**Comment by superboyiii** [comment]
@Crypto2 @shoxive Seems this issue is fixed in neo-cli-2.10.2, could you make a test? It seems work in mine.

**Comment by lock9** [comment]
Closing issue due to inactivity. @Crypto2 @shoxive please open a new issue if the problem persists. Thanks
